syntax/gomod.vim

```diff
@@ -45,6 +48,16 @@ highlight default link gomodString  String
 syntax match gomodReplaceOperator "\v\=\>"
 highlight default link gomodReplaceOperator Operator
 
+" match toolchain versions, per https://go.dev/doc/toolchain#version, e.g.
+"
+"  * go1.X.Y
+"  * go1.X
+"  * go1.XrcN
+"  * go1.X.Y-somesuffix
+"  * go1.XrcN-somesuffix more stuff
+syntax match gomodToolchainVersion "go1\.\d\+\(\(.\d\+\)\|\(rc\d\+\)\)\?\([ \t-].*\)\?" contained
```

Comment:
> I don't see any comments from you besides this one

Oof. I must have forgotten to hit `Start a review` before I submitted. I'll repost the quick summary here:

Two things about this after comparing this expression to the docs.
1. From the docs, it doesn't appear that any space separated strings are valid. Did you find something otherwise?
2. The docs indicate that a string is valid, but when I check rules, I think the only string that's allowed is `default`.

Reply:
Here's a link to the regular expression used by Go to validate the toolchain directive: https://cs.opensource.google/go/x/mod/+/refs/tags/v0.14.0:modfile/rule.go;l=309-311

Reply:
> From the docs, it doesn't appear that any space separated strings are valid. Did you find something otherwise?

I found some more detail digging around the stdlib https://go.googlesource.com/go/+/165383381199de6632665f43561e8e0dfc96f067/src/cmd/go/internal/gover/toolchain.go#24 (from looking for the source of the error `'go: invalid toolchain "%s" in go.mod'`

> Here's a link to the regular expression used by Go to validate the toolchain directive: https://cs.opensource.google/go/x/mod/+/refs/tags/v0.14.0:modfile/rule.go;l=309-311

More than happy to just use that since it's 1. simpler, 2. public documentation. It will permit versions that aren't accepted by the Go tool, e.g. `toolchain go1.2.2rc4` will satisfy that but error `go: invalid toolchain "go1.2.2rc4" in go.mod` if actually used. I think it's fine if the highlighting accepts some invalid values (as long as it doesn't reject any valid ones)

Reply:
Trying this with [f4260d8](https://github.com/fatih/vim-go/pull/3633/commits/f4260d8dacbf1398f09f924d5331265244e49727). It's still more restrictive than just matching on `^go1` (trying to follow the examples) let me know if you'd like it relaxed further

Reply:
No need to relax it further, and I think it may be appropriate to tighten it up so that we don't match invalid versions. 

This should also allow `default` if I'm not mistaken.

Reply:
@matthewhughes934 please feel free to let me know if you'd like to be done with this. I'm happy to take it as is and submit the incremental improvements on top of it myself.

Reply:
Good find in stdlib. It looks like my assertion about spaces was wrong: https://go.googlesource.com/go/+/165383381199de6632665f43561e8e0dfc96f067/src/cmd/go/internal/gover/toolchain.go#39.

With that in mind, I think your original regular expression was mostly right except for we should also allow `default`.

If I'm reading the stdlib source code correctly, then it looks like anything other than `default` is validated by `"internal/gover".IsValid`: https://go.googlesource.com/go/+/165383381199de6632665f43561e8e0dfc96f067/src/internal/gover/gover.go#92

Reply:
This is a fun tidbit from https://go.dev/doc/toolchain:
> The standard Go toolchains are named goV where V is a Go version denoting a beta release, release candidate, or release. For example, go1.21rc1 and go1.21.0 are toolchain names; go1.21 and go1.22 are not (the initial releases are go1.21.0 and go1.22.0), but go1.20 and go1.19 are.

Reply:
> With that in mind, I think your original regular expression was mostly right except for we should also allow default.

:+1: I've done this, and squashed all the changes down into [26f4ff5](https://github.com/fatih/vim-go/pull/3633/commits/26f4ff5569a3ba8a84c0be58fd5866d57471e492) there's some reasoning+a summary of this discussion in the commit

> please feel free to let me know if you'd like to be done with this. I'm happy to take it as is and submit the incremental improvements on top of it myself.

Go for it